extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -85,3 +88,343 @@ export class ISignatureVerifier {
+        // We don't have this key and we already tried cachebusting.
+        return VerificationStatus.ERROR_KEY_NOT_FOUND;
+      } else {
+        return keyPromise.then(key => {
+          if (!key) {
+            // This particular public key couldn't be imported. Probably the
+            // signing service's fault.
+            return VerificationStatus.UNVERIFIED;
+          }
+          const crypto = Services.cryptoFor(this.win_);
+          const startTime = this.getNow_();
+          return crypto.verifyPkcs(key, signature, creative).then(
+              result => {
+                const endTime = this.getNow_();
+                if (result) {
+                  lifecycleCallback('signatureVerifySuccess', {
+                    'met.delta.AD_SLOT_ID': endTime - startTime,
+                    'signingServiceName.AD_SLOT_ID': signingServiceName,
+                  });
+                  return VerificationStatus.OK;
+                } else {
+                  return VerificationStatus.ERROR_SIGNATURE_MISMATCH;
+                }
+              },
+              err => {
+                // Web Cryptography rejected the verification attempt. This
+                // hopefully won't happen in the wild, but browsers can be weird
+                // about this, so we need to guard against the possibility.
+                // Phone home to the AMP Project so that we can understand why
+                // this occurred.
+                const message = err && err.message;
+                dev().error(
+                    'AMP-A4A', `Failed to verify signature: ${message}`);
+                return VerificationStatus.UNVERIFIED;
+              });
+        });
+      }
+    });
+  }
+
+  /**
+   * Try to download the keyset for the named signing service and add a promise
+   * for each key to the `keys` object.
+   *
+   * @param {!Object<string, ?Promise<?webCrypto.CryptoKey>>} keys the object to
+   *     add each key promise to. This is mutated while the returned promise is
+   *     pending.
+   * @param {string} signingServiceName
+   * @param {?string} keypairId the keypair ID to include in the query string
+   *     for cachebusting purposes, or `null` if no cachebusting is needed
+   * @return {!Promise<boolean>} resolves after the mutation of `keys` is
+   *     complete, to `true` if the keyset was downloaded and parsed
+   *     successfully (even if some keys were malformed), or `false` if a
+   *     keyset-level failure occurred
+   * @private
+   */
+  fetchAndAddKeys_(keys, signingServiceName, keypairId) {
+    // TODO(@erwinmombay, #11081): eslint-disable-line needed due to linter bug
+    let url = this.signingServerURLs_[signingServiceName];
+    if (keypairId != null) {
+      url += '?kid=' + encodeURIComponent(keypairId);
+    }
+    return Services.xhrFor(this.win_)
+        .fetchJson(
+            url, // eslint-disable-line indent
+            {mode: 'cors', method: 'GET', ampCors: false, credentials: 'omit'})
```

Comment:
We used to have a nice comment about why we do ampCors false... can it come back?